cpp/oneapi/dal/backend/primitives/objective_function/logloss_functors_dpc.cpp

```diff
@@ -0,0 +1,532 @@
+            const auto last = blocking.get_block_end_index(b);
+            const auto first = blocking.get_block_start_index(b);
+            const auto length = last - first;
+            ONEDAL_ASSERT(0l < length);
+            auto x_rows = data_accessor.pull(q_, { first, last }, sycl::usm::alloc::device);
+            auto x_nd = pr::ndarray<Float, 2>::wrap(x_rows, { length, p_ });
+            ndview<Float, 1> buffer_batch = buffer_.slice(first, length);
+            ndview<Float, 1> hess_batch = raw_hessian_.slice(first, length);
+
+            sycl::event event_xv =
+                gemv(q_, x_nd, vec, buffer_batch, Float(1), Float(0), last_iter_deps);
+            event_xv.wait_and_throw(); // Without this line gemv does not work correctly
+
+            constexpr sycl::multiplies<Float> kernel_mul{};
+            auto event_dxv =
+                element_wise(q_, kernel_mul, buffer_batch, hess_batch, buffer_batch, { event_xv });
+
+            auto fill_tmp_event = fill<Float>(q_, tmp_ndview, Float(0), last_iter_deps);
+
+            sycl::event event_xtdxv = gemv(q_,
+                                           x_nd.t(),
+                                           buffer_batch,
+                                           tmp_ndview,
+                                           Float(1),
+                                           Float(0),
+                                           { event_dxv, fill_tmp_event });
+            event_xtdxv.wait_and_throw(); // Without this line gemv does not work correctly
+
+            sycl::event update_grad_e =
+                element_wise(q_, sycl::plus<>(), out, tmp_ndview, out, { event_xtdxv });
+            last_iter_deps = { update_grad_e };
+        }
+    }
+
+    if (comm_.get_rank_count() > 1) {
+        {
+            ONEDAL_PROFILER_TASK(hessp_allreduce);
+            auto hessp_arr = dal::array<Float>::wrap(q_,
+                                                     out.get_mutable_data(),
+                                                     out.get_count(),
+                                                     last_iter_deps);
+            comm_.allreduce(hessp_arr).wait();
+        }
+    }
+
+    const Float regularization_factor = L2_;
+
+    const auto kernel_regularization = [=](const Float a, const Float param) {
+        return a + param * regularization_factor;
+    };
+
+    auto add_regularization_event =
+        element_wise(q_, kernel_regularization, out, vec, out, last_iter_deps);
+
+    return add_regularization_event;
+}
+
+template <typename Float>
+sycl::event logloss_hessian_product<Float>::operator()(const ndview<Float, 1>& vec,
+                                                       ndview<Float, 1>& out,
+                                                       const event_vector& deps) {
+    if (fit_intercept_) {
+        return compute_with_fit_intercept(vec, out, deps);
+    }
+    else {
+        return compute_without_fit_intercept(vec, out, deps);
```

Comment:
Looks like these functions are pretty similar. You can reuse the code by atomizing functions